File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataframe/OrdersRevenueCaseWhen.scala
package io.devorbit.sparklabs.batch.sales.dataframe

import org.apache.spark.sql.functions._

object OrdersRevenueCaseWhen extends App {

  Utils.orderItems
    .join(Utils.orders, Seq("order_id"), "inner")
    .join(Utils.products, Seq("product_id"), "inner")
    .join(Utils.promotions, Seq("category"), "left")
    .withColumn("gross_revenue", col("price") * col("quantity"))
    .withColumn(
      "discount_pct",
      when(
        to_date(col("order_date")).between(to_date(col("start_date")), to_date(col("end_date"))),
        coalesce(col("discount_pct"), lit(0))
      ).otherwise(lit(0))
    )
    .withColumn("discount_value", col("gross_revenue") * col("discount_pct") / 100)
    .withColumn("net_revenue", col("gross_revenue") - col("discount_value"))
    .groupBy(col("order_id"))
    .agg(
      round(sum("gross_revenue"), 2).as("gross_revenue"),
      round(sum("discount_value"), 2).as("discount_value"),
      round(sum("net_revenue"), 2).as("net_revenue")
    )
    .orderBy(desc("net_revenue"))
    .show(false)

//  Thread.sleep(1000000)
}

File: python/sparklabs/batch/sales/dataset/orders_revenue_case_when.py
from pyspark.sql import functions as F

from sparklabs.batch.sales.dataset import utils


def run() -> None:
    (
        utils.order_items.join(utils.orders, ["order_id"], "inner")
        .join(utils.products, ["product_id"], "inner")
        .join(utils.promotions, ["category"], "left")
        .withColumn("gross_revenue", F.col("price") * F.col("quantity"))
        .withColumn(
            "discount_pct",
            F.when(
                F.to_date(F.col("order_date")).between(
                    F.to_date(F.col("start_date")), F.to_date(F.col("end_date"))
                ),
                F.coalesce(F.col("discount_pct"), F.lit(0)),
            ).otherwise(F.lit(0)),
        )
        .withColumn(
            "discount_value", F.col("gross_revenue") * F.col("discount_pct") / 100
        )
        .withColumn("net_revenue", F.col("gross_revenue") - F.col("discount_value"))
        .groupBy(F.col("order_id"))
        .agg(
            F.round(F.sum("gross_revenue"), 2).alias("gross_revenue"),
            F.round(F.sum("discount_value"), 2).alias("discount_value"),
            F.round(F.sum("net_revenue"), 2).alias("net_revenue"),
        )
        .orderBy(F.desc("net_revenue"))
        .show(truncate=False)
    )


if __name__ == "__main__":
    run()
